### PR TITLE
feat(azure): Checks related to Azure Keyvault

### PR DESCRIPTION
**Pull request by pedrooot**
### Context
This pr adds 8 new checks related to Key Vault service in Azure:

- [x] keyvault_key_expiration_set_for_rbac_keyvault
- [x] keyvault_key_expiration_set_for_non_rbac_keyvault
- [x] keyvault_secret_expiration_set_for_rbac_keyvault
- [x] keyvault_secret_expiration_set_for_non_rbac_keyvault
- [x] keyvault_recoverable
- [x] keyvault_rbac_enabled
- [x] keyvault_private_endpoints
- [x] keyvault_key_rotation_enabled

### Description

Please include a summary of the change and which issue is fixed. List any dependencies that are required for this change.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: poetry.lock [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# This file is automatically @generated by Poetry 1.7.1 and should not be changed by hand.
+# This file is automatically @generated by Poetry 1.7.0 and should not be changed by hand.
```
Comment (MrCloudSec) [inline poetry.lock]:
```suggestion
# This file is automatically @generated by Poetry 1.7.1 and should not be changed by hand.
```

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.metadata.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  "SubServiceName": "",
+  "ResourceIdTemplate": "",
+  "Severity": "high",
+  "ResourceType": "KeyVault",
+  "Description": "Ensure that all Keys in Non Role Based Access Control (RBAC) Azure Key Vaults have an expiration date set.",
+  "Risk": "Azure Key Vault enables users to store and use cryptographic keys within the Microsoft Azure environment. The exp (expiration date) attribute identifies the expiration date on or after which the key MUST NOT be used for a cryptographic operation. By default, keys never expire. It is thus recommended that keys be rotated in the key vault and set an explicit expiration date for all keys. This ensures that the keys cannot be used beyond their assigned lifetimes.",
+  "RelatedUrl": "https://docs.microsoft.com/en-us/azure/key-vault/key-vault-whatis",
+  "Remediation": {
+    "Code": {
+      "CLI": "https://www.trendmicro.com/cloudoneconformity/knowledge-base/azure/KeyVault/key-expiration-check.html#",
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.metadata.json]:
```suggestion
      "CLI": "https://docs.bridgecrew.io/docs/set-an-expiration-date-on-all-keys#cli-command",
```

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.metadata.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
+  "Provider": "azure",
+  "CheckID": "keyvault_key_expiration_set_for_non_rbac_keyvault",
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.metadata.json]:
```suggestion
  "CheckID": "keyvault_non_rbac_key_expiration_set",
```

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+                    report = Check_Report_Azure(self.metadata())
+                    report.subscription = subscription
+                    report.resource_name = keyvault.name
+                    report.resource_id = keyvault.id
+                    report.status = "PASS"
+                    report.status_extended = f"Keyvault {keyvault.name} from subscription {subscription} has all the keys with expiration date set."
+                    for key in keyvault.keys:
+                        if not key.attributes.expires and key.attributes.enabled:
+                            report.status = "FAIL"
+                            report.status_extended = f"Keyvault {keyvault.name} from subscription {subscription} has a key without expiration date set."
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.py]: Make a FAIL for each misconfigured key.

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.metadata.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
+{
+  "Provider": "azure",
+  "CheckID": "keyvault_key_expiration_set_for_rbac_keyvault",
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.metadata.json]:
```suggestion
  "CheckID": "keyvault_rbac_key_expiration_set",
```

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.metadata.json [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  "SubServiceName": "",
+  "ResourceIdTemplate": "",
+  "Severity": "high",
+  "ResourceType": "KeyVault",
+  "Description": "Ensure that all Keys in Role Based Access Control (RBAC) Azure Key Vaults have an expiration date set",
+  "Risk": "Azure Key Vault enables users to store and use cryptographic keys within the Microsoft Azure environment. The exp (expiration date) attribute identifies the expiration date on or after which the key MUST NOT be used for encryption of new data, wrapping of new keys, and signing. By default, keys never expire. It is thus recommended that keys be rotated in the key vault and set an explicit expiration date for all keys to help enforce the key rotation. This ensures that the keys cannot be used beyond their assigned lifetimes.",
+  "RelatedUrl": "https://docs.microsoft.com/en-us/azure/key-vault/key-vault-whatis",
+  "Remediation": {
+    "Code": {
+      "CLI": "https://www.trendmicro.com/cloudoneconformity/knowledge-base/azure/KeyVault/key-expiration-check.html#",
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.metadata.json]:
```suggestion
      "CLI": "https://docs.bridgecrew.io/docs/set-an-expiration-date-on-all-keys#cli-command",
```

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+                    report.subscription = subscription
+                    report.resource_name = keyvault.name
+                    report.resource_id = keyvault.id
+                    report.status = "PASS"
+                    report.status_extended = f"Keyvault {keyvault.name} from subscription {subscription} has all the keys with expiration date set."
+                    for key in keyvault.keys:
+                        if not key.attributes.expires and key.attributes.enabled:
+                            report.status = "FAIL"
+                            report.status_extended = f"Keyvault {keyvault.name} from subscription {subscription} has a key without expiration date set."
+                            break
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.py]: Same as the other check.

File: prowler/providers/azure/services/keyvault/keyvault_private_endpoints/keyvault_private_endpoints.metadata.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  "SubServiceName": "",
+  "ResourceIdTemplate": "",
+  "Severity": "high",
+  "ResourceType": "KeyVault",
+  "Description": "Private endpoints will secure network traffic from Azure Key Vault to the resources requesting secrets and keys.",
+  "Risk": "Private endpoints will keep network requests to Azure Key Vault limited to the endpoints attached to the resources that are whitelisted to communicate with each other. Assigning the Key Vault to a network without an endpoint will allow other resources on that network to view all traffic from the Key Vault to its destination. In spite of the complexity in configuration, this is recommended for high security secrets.",
+  "RelatedUrl": "https://docs.microsoft.com/en-us/azure/private-link/private-endpoint-overview",
+  "Remediation": {
+    "Code": {
+      "CLI": "az keyvault private-endpoint-connection show -g <resourceGroup> --vault-name <keyVaultName>",
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_private_endpoints/keyvault_private_endpoints.metadata.json]:
```suggestion
      "CLI": "",
```

File: prowler/providers/azure/services/keyvault/keyvault_private_endpoints/keyvault_private_endpoints.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+                report.resource_name = keyvault.name
+                report.resource_id = keyvault.id
+                report.status = "FAIL"
+                report.status_extended = f"Keyvault {keyvault.name} from subscription {subscription} is not using private endpoints."
+                if keyvault.properties.private_endpoint_connections:
+                    print(
+                        keyvault.properties.private_endpoint_connections[
+                            0
+                        ].__class__.__name__
+                    )
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_private_endpoints/keyvault_private_endpoints.py]:
```suggestion
```

File: prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault.metadata.json [hunk 10]
```diff
@@ -0,0 +1,30 @@
+{
+  "Provider": "azure",
+  "CheckID": "keyvault_secret_expiration_set_for_non_rbac_keyvault",
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault.metadata.json]:
```suggestion
  "CheckID": "keyvault_non_rbac_secret_expiration_set",
```

File: prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+                    report.subscription = subscription
+                    report.resource_name = keyvault.name
+                    report.resource_id = keyvault.id
+                    report.status = "PASS"
+                    report.status_extended = f"Keyvault {keyvault.name} from subscription {subscription} has all the secrets with expiration date set."
+                    for secret in keyvault.secrets:
+                        if not secret.attributes.expires and secret.attributes.enabled:
+                            report.status = "FAIL"
+                            report.status_extended = f"Keyvault {keyvault.name} from subscription {subscription} has a secret without expiration date set."
+                            break
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault.py]: Make a FAIL for each secret please.

File: prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_rbac_keyvault/keyvault_secret_expiration_set_for_rbac_keyvault.metadata.json [hunk 12]
```diff
@@ -0,0 +1,30 @@
+{
+  "Provider": "azure",
+  "CheckID": "keyvault_secret_expiration_set_for_rbac_keyvault",
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_rbac_keyvault/keyvault_secret_expiration_set_for_rbac_keyvault.metadata.json]:
```suggestion
  "CheckID": "keyvault_rbac_secret_expiration_set",
```

File: prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_rbac_keyvault/keyvault_secret_expiration_set_for_rbac_keyvault.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+                    report.subscription = subscription
+                    report.resource_name = keyvault.name
+                    report.resource_id = keyvault.id
+                    report.status = "PASS"
+                    report.status_extended = f"Keyvault {keyvault.name} from subscription {subscription} has all the secrets with expiration date set."
+                    for secret in keyvault.secrets:
+                        if not secret.attributes.expires and secret.attributes.enabled:
+                            report.status = "FAIL"
+                            report.status_extended = f"Keyvault {keyvault.name} from subscription {subscription} has a secret without expiration date set."
+                            break
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_rbac_keyvault/keyvault_secret_expiration_set_for_rbac_keyvault.py]: Make a FAIL for each secret please.

**Review by MrCloudSec — CHANGES_REQUESTED** [review]
Please, review my comments.

File: poetry.lock (hunk 1 above)
Comment (pedrooot) [reply poetry.lock]: Done

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.metadata.json (hunk 2 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.metadata.json]: Done

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.metadata.json (hunk 3 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.metadata.json]: Done

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.py (hunk 4 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault/keyvault_key_expiration_set_for_non_rbac_keyvault.py]: Done

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.metadata.json (hunk 5 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.metadata.json]: Done

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.metadata.json (hunk 6 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.metadata.json]: Done

File: prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.py (hunk 7 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_key_expiration_set_for_rbac_keyvault/keyvault_key_expiration_set_for_rbac_keyvault.py]: Done

File: prowler/providers/azure/services/keyvault/keyvault_private_endpoints/keyvault_private_endpoints.py (hunk 9 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_private_endpoints/keyvault_private_endpoints.py]: Done

File: prowler/providers/azure/services/keyvault/keyvault_private_endpoints/keyvault_private_endpoints.metadata.json (hunk 8 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_private_endpoints/keyvault_private_endpoints.metadata.json]: Done

File: prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault.metadata.json (hunk 10 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault.metadata.json]: Done

File: prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault.py (hunk 11 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault/keyvault_secret_expiration_set_for_non_rbac_keyvault.py]: Done

File: prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_rbac_keyvault/keyvault_secret_expiration_set_for_rbac_keyvault.metadata.json (hunk 12 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_rbac_keyvault/keyvault_secret_expiration_set_for_rbac_keyvault.metadata.json]: Done

File: prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_rbac_keyvault/keyvault_secret_expiration_set_for_rbac_keyvault.py (hunk 13 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_secret_expiration_set_for_rbac_keyvault/keyvault_secret_expiration_set_for_rbac_keyvault.py]: Done

File: prowler/providers/azure/services/keyvault/keyvault_service.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+            keys_list = client.keys.list(resource_group, keyvault_name)
+            for key in keys_list:
+                keys.append(
+                    Key(
+                        id=getattr(key, "id", None),
+                        name=getattr(key, "name", None),
+                        enabled=getattr(key.attributes, "enabled", None),
+                        location=getattr(key, "location", None),
+                        attributes=getattr(key, "attributes", None),
+                        rotation_policy=None,
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_service.py]:
```suggestion
```
This is not needed.

File: prowler/providers/azure/services/keyvault/keyvault_service.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+            client = self.clients[subscription]
+            keys_list = client.keys.list(resource_group, keyvault_name)
+            for key in keys_list:
+                keys.append(
+                    Key(
+                        id=getattr(key, "id", None),
+                        name=getattr(key, "name", None),
+                        enabled=getattr(key.attributes, "enabled", None),
+                        location=getattr(key, "location", None),
+                        attributes=getattr(key, "attributes", None),
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_service.py]: Are we handling the cases when these attributes are None?

File: prowler/providers/azure/services/keyvault/keyvault_key_rotation_enabled/keyvault_key_rotation_enabled.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+                    report = Check_Report_Azure(self.metadata())
+                    report.subscription = subscription
+                    report.resource_name = keyvault.name
+                    report.resource_id = keyvault.id
+                    for key in keyvault.keys:
+                        if key.rotation_policy:
+                            if (
+                                key.rotation_policy.lifetime_actions[0].action
+                                == "Rotate"
+                            ):
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_key_rotation_enabled/keyvault_key_rotation_enabled.py]: There is a missing else here

Comment (MrCloudSec) [reply prowler/providers/azure/services/keyvault/keyvault_key_rotation_enabled/keyvault_key_rotation_enabled.py]: Join the two IFs above

File: prowler/providers/azure/services/keyvault/keyvault_service.py (hunk 14 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_service.py]: Done!

File: prowler/providers/azure/services/keyvault/keyvault_key_rotation_enabled/keyvault_key_rotation_enabled.py (hunk 16 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_key_rotation_enabled/keyvault_key_rotation_enabled.py]: Done!

File: prowler/providers/azure/services/keyvault/keyvault_service.py (hunk 15 above)
Comment (pedrooot) [reply prowler/providers/azure/services/keyvault/keyvault_service.py]: Done! Good catch

File: prowler/providers/azure/services/keyvault/keyvault_key_rotation_enabled/keyvault_key_rotation_enabled.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+            for keyvault in key_vaults:
+                if keyvault.keys:
+                    report = Check_Report_Azure(self.metadata())
+                    report.subscription = subscription
+                    report.resource_name = keyvault.name
+                    report.resource_id = keyvault.id
+                    for key in keyvault.keys:
+                        if (
+                            key.rotation_policy
+                            and key.rotation_policy.lifetime_actions[0].action
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_key_rotation_enabled/keyvault_key_rotation_enabled.py:19]:
```suggestion
                            and key.rotation_policy.lifetime_actions
                            and key.rotation_policy.lifetime_actions[0].action
```

File: prowler/providers/azure/services/keyvault/keyvault_service.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+        secrets = []
+        try:
+            client = self.clients[subscription]
+            secrets_list = client.secrets.list(resource_group, keyvault_name)
+            for secret in secrets_list:
+                secrets.append(
+                    Secret(
+                        id=getattr(secret, "id", ""),
+                        name=getattr(secret, "name", ""),
+                        enabled=getattr(secret.properties.attributes, "enabled", ""),
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/keyvault/keyvault_service.py]:
```suggestion
                        enabled=getattr(secret.properties.attributes, "enabled", False),
```

**Review by MrCloudSec — CHANGES_REQUESTED** [review]
Review my comments please

**Review by MrCloudSec — APPROVED** [review]
Awesome job!

**Pull request merged**